// doxjs/js/DoxArray.js
import { defineProperty, getKeys, isArray, isObject, isPrimary, isSame } from "./utils";
import { DoxObject } from "./DoxObject";
import * as dfa from "./core";
export class DoxArray {
    constructor(target, root, parentEvents = []) {
        this.runtimeEvent = {
            isLegal: false,
            value: (store) => { }
        };
        this.eventsStore = {};
        this.dataToProxy = [];
        this.proxy = {
            length: 0
        };
        this.delicate = true;
        this.parentEvents = [];
        this.target = target;
        this.root = root;
        this.keys = getKeys(target);
        this.parentEvents = parentEvents;
        this.init();
    }
    init() {
        this.keys.forEach(key => {
            this.eventsStore[key] = [];
            this.proxy[key] = undefined;
            defineProperty(this.proxy, key, {
                get: () => {
                    if (this.runtimeEvent.isLegal) {
                        this.eventsStore[key] = this.eventsStore[key].reduce((i, f) => {
                            if (f !== i[0]) {
                                i.push(f);
                            }
                            return i;
                        }, [this.runtimeEvent.value]);
                    }
                    if ((this.dataToProxy[key] instanceof DoxObject) || (this.dataToProxy[key] instanceof DoxArray)) {
                        if (this.runtimeEvent.isLegal) {
                            this.dataToProxy[key].extendsRuntimeEvent(this.runtimeEvent);
                        }
                        return this.dataToProxy[key].observe();
                    }
                    else if (isPrimary(this.dataToProxy[key])) {
                        return this.dataToProxy[key];
                    }
                },
                set: (newValue) => {
                    if (!isSame(newValue, this.dataToProxy[key])) {
                        if (isPrimary(newValue)) {
                            this.dataToProxy[key] = newValue;
                        }
                        else if (isObject(newValue)) {
                            this.dataToProxy[key] = new DoxObject(newValue, this.root);
                        }
                        else if (isArray(newValue)) {
                            this.dataToProxy[key] = new DoxArray(newValue, this.root, this.parentEvents);
                        }
                    }
                    if (this.delicate) {
                        this.eventsStore[key].forEach(f => f(this.root.observe()));
                    }
                }
            });
            this.proxy[key] = this.target[key];
        });
        defineProperty(this.proxy, "length", {
            enumerable: false,
            configurable: false,
            writable: true,
            value: this.target.length
        });
        dfa.methods.forEach(method => {
            defineProperty(this.proxy, method, {
                configurable: false,
                writable: false,
                enumerable: false,
                value: (...args) => {
                    this.toggleDelicate(false);
                    let result = dfa[method].call(this.proxy, ...args);
                    this.parentEvents.forEach(f => {
                        f(this.root.observe());
                    });
                    this.toggleDelicate();
                    return result;
                }
            });
        });
    }
    toggleDelicate(is = true) {
        this.delicate = is;
    }
    extendsRuntimeEvent(callback) {
        this.runtimeEvent = callback;
    }
    observe() {
        return this.proxy;
    }
}
