// doxjs/src/index.ts
import { DoxObject } from "./DoxObject";
import { isObject, warn, getKeys, hasProperty, _toString, isFunction } from "./utils";
import { RuntimeEvent, DoxActions, DoxCallback } from "./DoxInterfaces";

export default class DoxJS<T> {
    private runtimeEvent: RuntimeEvent<T>;
    private sourceObject: DoxObject<T>;
    private actions: DoxActions;
    private listeners: {
        [key: string]: DoxCallback<T>
    };

    constructor(target: T) {
        if (!isObject(target)) {
            warn(`DoxJS only accept object as argument`);
        } else {
            this.sourceObject = new DoxObject(target, this);
        }
    }

    public subscribe<C>(action: DoxCallback<T> | string, excuteLater: boolean = false, context?: C) {
        if (isFunction(action)) {
            this.runtimeEvent = {
                isLegal: true,
                value: <DoxCallback<T>>action
            };

            let fn = () => {
                this.sourceObject.extendsRuntimeEvent(this.runtimeEvent);
                (<DoxCallback<T>>action)(this.sourceObject.observe());
                this.runtimeEvent.isLegal = false;
            }

            return (excuteLater ? fn : fn());
        } else if(_toString.call(action) === '[object String]') {
            if (hasProperty(this.listeners, action)) {
                if (context) {
                    this.subscribe(this.listeners[<string>action].bind(context), excuteLater);
                } else {
                    this.subscribe(this.listeners[<string>action], excuteLater);
                }
            }
        } else {
            warn(`First parameter of subscribe should be Function or String`);
        }
    }

    public bindListeners(listeners: { [key: string]: DoxCallback<T> } = {}) {
        this.listeners = listeners;
    }

    public observe(): T {
        return this.sourceObject.observe();
    }

    public bindActions(actions: DoxActions = {}) {
        this.actions = actions;
    }

    public dispatch(action: string, ...args: any[]) {
        if (hasProperty(this.actions, action)) {
            this.actions[action](this.sourceObject.observe(), ...args);
        }
    }
}
